test/boost/alternator_expressions_test.cc

```
@@ -0,0 +1,29 @@
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#include <boost/test/unit_test.hpp>
+#include <seastar/testing/test_case.hh>
+#include <seastar/core/coroutine.hh>
+#include "alternator/expressions.hh"
+
+// ANTLR3 leaks memory when it tries to recover from missing token.
+// Reproduces #25878
+SEASTAR_TEST_CASE(missing_tokens_memory_leak) {
+    BOOST_REQUIRE_THROW(alternator::parse_update_expression("SET a :v"), alternator::expressions_syntax_error); // missing '='
+    BOOST_REQUIRE_THROW(alternator::parse_update_expression("DELETE a v"), alternator::expressions_syntax_error); // missing ':'
+    BOOST_REQUIRE_THROW(alternator::parse_update_expression("ADD a v"), alternator::expressions_syntax_error); // missing ':'
+    
+    BOOST_REQUIRE_THROW(alternator::parse_condition_expression("size(a < 5", "Test"), alternator::expressions_syntax_error); // missing ')'
+    BOOST_REQUIRE_THROW(alternator::parse_condition_expression("a IN :x)", "Test"), alternator::expressions_syntax_error); // missing '('
+    BOOST_REQUIRE_THROW(alternator::parse_condition_expression("a IN (:x", "Test"), alternator::expressions_syntax_error); // missing ')'
+    BOOST_REQUIRE_THROW(alternator::parse_condition_expression("a BETWEEN :x AN :y", "Test"), alternator::expressions_syntax_error); // missing 'AND'
+    BOOST_REQUIRE_THROW(alternator::parse_condition_expression("a BETWEEN :x :y", "Test"), alternator::expressions_syntax_error); // missing 'AND'
+
+    BOOST_REQUIRE_THROW(alternator::parse_projection_expression("a[0.b"), alternator::expressions_syntax_error); // missing ']'
+    co_return;
```

A thought: Can you please try to add another test function (it's not a "missing token" issue) for parsing `a[99999999999999999]`? I want to see that despite the throw in stoi, we still report the error correctly and don't leak. Or - if it does leak, we'll have a reproducer for a new issue ;-)

---

I checked and it reports std::out_of_range and doesn't leak.
But I will add separate issue for that case, as I think it is still not correct.
I believe it should be ignored, as comment in python tests says:
```
    # If some of the paths are not available, they are silently ignored (just
    # like they returned an empty item when used alone earlier)
```
I think it should be covered in python tests, as we have such tests there.
Unless python test can't detect memory leaks, but in that case we may want to add more cases.

---

Well, DynamoDB also has some limit - it just gives better message.
Nevertheless it is a candidate for python test.

I will add also here, with other cases that trigger exception.

---

Very interesting. I agree with your view that if `a[999]` is supposed to return nothing because the list doesn't have 999 items, then certainly `a[999999999999999]` can just return nothing - and not fail.

However, it's still customary for parsers to parse first and only then check for other errors, so it's not surprising for the parsing of 999999999999999 to fail early before realizing it could have just done nothing. Let's try to do the same thing that DynamoDB does in this error, unless we have a reason not to. You're right that this needs a separate test in Python, and is not really related to this patch. I can do that.